File: data_structure/graph/src/GraphDemo.py
# Python
# データ構造: グラフ (Graph)

class GraphData:
    def __init__(self):
        # 隣接ノードとその辺の重みを格納します。
        # キーは頂点、値はその頂点に隣接する頂点と重みの辞書です。
        # 例: { 'A': {'B': 1, 'C': 4}, 'B': {'A': 1, 'C': 2, 'D': 5}, ... }
        self._data = {}

    def get(self):
        # グラフの内部データを取得します。
        return self._data

    def get_vertices(self):
        # グラフの全頂点をリストとして返します。
        return list(self._data.keys())
    
    def get_edges(self):
        # グラフの全辺をリストとして返します。
        # 無向グラフの場合、(u, v, weight) の形式で返します。
        # 重複を避けるためにセットを使用します。
        edges = set()
        for vertex in self._data:
            for neighbor, weight in self._data[vertex]:
                # 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                edge = tuple(sorted((vertex, neighbor)))
                edges.add((edge[0], edge[1], weight)) # (u, v, weight) の形式で格納
        return list(edges)
    
    def get_neighbors(self, vertex):
        # 指定された頂点の隣接ノードと辺の重みのリストを返します。
        # 形式: [(隣接頂点, 重み), ...]
        if vertex in self._data:
            return self._data[vertex]
        else:
            return None # 頂点が存在しない場合はNoneを返す

    def get_edge_weight(self, vertex1, vertex2):
        # 指定された2つの頂点間の辺の重みを返します。
        # 辺が存在しない場合はNoneを返します。
        if vertex1 in self._data and vertex2 in self._data:
            for neighbor, weight in self._data[vertex1]:
                if neighbor == vertex2:
                    return weight
        return None # 辺が存在しない場合

    def get_vertice(self, vertex):
        # 頂点がグラフに存在するか確認する
        if vertex in self._data:
            # 存在する場合は、その頂点の隣接リスト（関連する値）を返す
            return self._data[vertex]
        else:
            # 存在しない場合はメッセージを表示し、Noneを返す
            print(f"ERROR: {vertex}は範囲外です")
            return None

    def get_edge(self, vertex1, vertex2):
        # 指定された2つの頂点間に辺が存在するかを確認する
        # 両方の頂点がグラフに存在する必要がある
        if vertex1 in self._data and vertex2 in self._data:
            # vertex1の隣接リストにvertex2が含まれているかを確認
            # 無向グラフなので、片方を確認すれば十分
            return vertex2 in self._data[vertex1]
        else:
            # どちらかの頂点が存在しない場合は辺も存在しない
            return False

    def add_vertex(self, vertex):
        # 新しい頂点をグラフに追加します。
        if vertex not in self._data:
            self._data[vertex] = []
            return True
        # 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return True
    
    def add_edge(self, vertex1, vertex2, weight):
        # 両頂点間に辺を追加します。重みを指定します。
        # 頂点がグラフに存在しない場合は追加します。
        if vertex1 not in self._data:
            self.add_vertex(vertex1)
        if vertex2 not in self._data:
            self.add_vertex(vertex2)
        
        # 両方向に辺を追加する（無向グラフ）
        # 既に同じ辺が存在する場合は重みを更新するかどうか？
        # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
        # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
        
        # vertex1 -> vertex2 の辺を追加（重み付き）
        edge_exists_v1v2 = False
        for i, (neighbor, _) in enumerate(self._data[vertex1]):
            if neighbor == vertex2:
                self._data[vertex1][i] = (vertex2, weight) # 既に存在する場合は重みを更新
                edge_exists_v1v2 = True
                break
        if not edge_exists_v1v2:
            self._data[vertex1].append((vertex2, weight))

        # vertex2 -> vertex1 の辺を追加（重み付き）
        edge_exists_v2v1 = False
        for i, (neighbor, _) in enumerate(self._data[vertex2]):
            if neighbor == vertex1:
                self._data[vertex2][i] = (vertex1, weight) # 既に存在する場合は重みを更新
                edge_exists_v2v1 = True
                break
        if not edge_exists_v2v1:
            self._data[vertex2].append((vertex1, weight))
        
        return True
    
    def remove_vertex(self, vertex):
        # 頂点とそれに関連する辺を削除する
        if vertex in self._data:
            # この頂点への参照を他の頂点の隣接リストから削除する
            for v in self._data:
                if vertex in self._data[v]:
                    self._data[v].remove(vertex)
            # 頂点自体を削除する
            del self._data[vertex]
            return True
        else:
            print(f"ERROR: {vertex} は範囲外です")
            return False

    def remove_edge(self, vertex1, vertex2):
        # 両頂点間の辺を削除します。
        if vertex1 in self._data and vertex2 in self._data:
            removed = False
            # vertex1 から vertex2 への辺を削除
            original_len_v1 = len(self._data[vertex1])
            self._data[vertex1] = [(neighbor, weight) for neighbor, weight in self._data[vertex1] if neighbor != vertex2]
            if len(self._data[vertex1]) < original_len_v1:
                removed = True

            # vertex2 から vertex1 への辺を削除
            original_len_v2 = len(self._data[vertex2])
            self._data[vertex2] = [(neighbor, weight) for neighbor, weight in self._data[vertex2] if neighbor != vertex1]
            if len(self._data[vertex2]) < original_len_v2:
                removed = True
                
            return removed # 少なくとも片方向が削除されたか

        else:
            print(f"ERROR: {vertex1} または {vertex2} は範囲外です")
            return False

    def is_empty(self):
        # グラフが空かどうか
        return len(self._data) == 0
    
    def size(self):
        # グラフの頂点数を返す
        return len(self._data)
    
    def clear(self):
        # グラフを空にする
        self._data = {}
        return True

def main():
    print("Graph TEST -----> start")

    print("\nnew")
    graph_data = GraphData()
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nis_empty")
    output = graph_data.is_empty()
    print(f"  出力値: {output}")

    print("\nsize")
    output = graph_data.size()
    print(f"  出力値: {output}")

    inputList = ['A', 'B', 'C']
    for input in inputList:
        print("\nadd_vertex")
        print(f"  入力値: {input}")
        output = graph_data.add_vertex(input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nget_vertices")
    output = graph_data.get_vertices()
    print(f"  出力値: {output}")

    print("\nsize")
    output = graph_data.size()
    print(f"  出力値: {output}")

    print("\nadd_edge")
    inputList = [('A', 'B', 4), ('B', 'C', 2), ('C', 'A', 3)]
    for input in inputList:
        print(f"  入力値: {input}")
        output = graph_data.add_edge(*input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nget_vertices")
    output = graph_data.get_vertices()
    print(f"  出力値: {output}")

    print("\nget_edges")
    output = graph_data.get_edges()
    print(f"  出力値: {output}")

    print("\nsize")
    output = graph_data.size()
    print(f"  出力値: {output}")

    print("\nget_vertice")
    input = 'B'
    print(f"  入力値: '{input}'")
    output = graph_data.get_vertice(input)
    print(f"  出力値: {output}")

    print("\nget_vertice")
    input = 'E'
    print(f"  入力値: '{input}'")
    output = graph_data.get_vertice(input)
    print(f"  出力値: {output}")

    print("\nremove_edge")
    input = ('A', 'B')
    print(f"  入力値: {input}")
    output = graph_data.remove_edge(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nremove_edge")
    input = ('A', 'C') 
    print(f"  入力値: {input}")
    output = graph_data.remove_edge(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nget_edges")
    output = graph_data.get_edges()
    print(f"  出力値: {output}")

    print("\nremove_vertex")
    input = 'B'
    print(f"  入力値: {input}")
    output = graph_data.remove_vertex(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nremove_vertex")
    input = 'Z'
    print(f"  入力値: {input}")
    output = graph_data.remove_vertex(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nget_vertices")
    output = graph_data.get_vertices()
    print(f"  出力値: {output}")

    print("\nget_edges")
    output = graph_data.get_edges()
    print(f"  出力値: {output}")

    print("\nsize")
    output = graph_data.size()
    print(f"  出力値: {output}")

    print("\nget_vertice")
    input = 'B'
    print(f"  入力値: {input}")
    output = graph_data.get_vertice(input)
    print(f"  出力値: {output}")

    print("\nclear")
    output = graph_data.clear()
    print(f"  出力値: {output}")

    print("\nis_empty")
    output = graph_data.is_empty()
    print(f"  出力値: {output}")

    print("\nsize")
    output = graph_data.size()
    print(f"  出力値: {output}")

    print("\nget_vertices")
    output = graph_data.get_vertices()
    print(f"  出力値: {output}")

    print("\nget_edges")
    output = graph_data.get_edges()
    print(f"  出力値: {output}")

    print("\nGraph TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/graph/src/GraphDemo.rb
# Ruby
# データ構造: グラフ (Graph)

class GraphData
  def initialize
    # Hash storing vertices and their adjacent vertices with weights
    # Format: { 'A' => [['B', 1], ['C', 4]], 'B' => [['A', 1], ['C', 2], ['D', 5]], ... }
    @data = {}
  end

  def get
    @data
  end

  def get_vertices
    @data.keys
  end
  
  def get_edges
    # Create a set of unique edges to avoid duplicates
    edges = Set.new
    @data.each do |vertex, neighbors|
      neighbors.each do |neighbor, weight|
        # Normalize the edge to prevent duplicates
        edge = [vertex, neighbor].sort
        edges.add([edge[0], edge[1], weight])
      end
    end
    edges.to_a
  end
  
  def get_neighbors(vertex)
    @data[vertex] || nil
  end

  def get_edge_weight(vertex1, vertex2)
    return nil unless @data.key?(vertex1) && @data.key?(vertex2)
    
    @data[vertex1].each do |neighbor, weight|
      return weight if neighbor == vertex2
    end
    nil
  end

  def get_vertice(vertex)
    if @data.key?(vertex)
      @data[vertex]
    else
      puts "ERROR: #{vertex} is out of range"
      nil
    end
  end

  def get_edge(vertex1, vertex2)
    return false unless @data.key?(vertex1) && @data.key?(vertex2)
    
    @data[vertex1].any? { |neighbor, _| neighbor == vertex2 }
  end

  def add_vertex(vertex)
    @data[vertex] = [] unless @data.key?(vertex)
    true
  end
  
  def add_edge(vertex1, vertex2, weight)
    add_vertex(vertex1) unless @data.key?(vertex1)
    add_vertex(vertex2) unless @data.key?(vertex2)
    
    # Add or update edge from vertex1 to vertex2
    edge_exists_v1v2 = @data[vertex1].any? do |neighbor, _|
      if neighbor == vertex2
        @data[vertex1][(@data[vertex1].index { |n, _ | n == vertex2 })] = [vertex2, weight]
        true
      end
    end
    @data[vertex1] << [vertex2, weight] unless edge_exists_v1v2

    # Add or update edge from vertex2 to vertex1
    edge_exists_v2v1 = @data[vertex2].any? do |neighbor, _|
      if neighbor == vertex1
        @data[vertex2][(@data[vertex2].index { |n, _ | n == vertex1 })] = [vertex1, weight]
        true
      end
    end
    @data[vertex2] << [vertex1, weight] unless edge_exists_v2v1
    
    true
  end
  
  def remove_vertex(vertex)
    if @data.key?(vertex)
      # Remove references to this vertex from other vertices
      @data.each_value do |neighbors|
        neighbors.reject! { |neighbor, _| neighbor == vertex }
      end
      # Remove the vertex itself
      @data.delete(vertex)
      true
    else
      puts "ERROR: #{vertex} is out of range"
      false
    end
  end

  def remove_edge(vertex1, vertex2)
    if @data.key?(vertex1) && @data.key?(vertex2)
      removed = false
      
      # Remove edge from vertex1 to vertex2
      original_len_v1 = @data[vertex1].length
      @data[vertex1].reject! { |neighbor, _| neighbor == vertex2 }
      removed = true if @data[vertex1].length < original_len_v1

      # Remove edge from vertex2 to vertex1
      original_len_v2 = @data[vertex2].length
      @data[vertex2].reject! { |neighbor, _| neighbor == vertex1 }
      removed = true if @data[vertex2].length < original_len_v2
      
      removed
    else
      puts "ERROR: #{vertex1} or #{vertex2} is out of range"
      false
    end
  end

  def is_empty?
    @data.empty?
  end
  
  def size
    @data.size
  end
  
  def clear
    @data.clear
    true
  end
end

def main
  puts "Graph TEST -----> start"

  puts "\nnew"
  graph_data = GraphData.new
  puts "  Current data: #{graph_data.get}"

  puts "\nis_empty?"
  output = graph_data.is_empty?
  puts "  Output value: #{output}"

  puts "\nsize"
  output = graph_data.size
  puts "  Output value: #{output}"

  input_list = ['A', 'B', 'C']
  input_list.each do |input|
    puts "\nadd_vertex"
    puts "  Input value: #{input}"
    output = graph_data.add_vertex(input)
    puts "  Output value: #{output}"
  end
  puts "  Current data: #{graph_data.get}"

  puts "\nget_vertices"
  output = graph_data.get_vertices
  puts "  Output value: #{output}"

  puts "\nsize"
  output = graph_data.size
  puts "  Output value: #{output}"

  puts "\nadd_edge"
  input_list = [['A', 'B', 4], ['B', 'C', 2], ['C', 'A', 3]]
  input_list.each do |input|
    puts "  Input value: #{input}"
    output = graph_data.add_edge(*input)
    puts "  Output value: #{output}"
  end
  puts "  Current data: #{graph_data.get}"

  puts "\nget_vertices"
  output = graph_data.get_vertices
  puts "  Output value: #{output}"

  puts "\nget_edges"
  output = graph_data.get_edges
  puts "  Output value: #{output}"

  puts "\nsize"
  output = graph_data.size
  puts "  Output value: #{output}"

  puts "\nget_vertice"
  input = 'B'
  puts "  Input value: '#{input}'"
  output = graph_data.get_vertice(input)
  puts "  Output value: #{output}"

  puts "\nget_vertice"
  input = 'E'
  puts "  Input value: '#{input}'"
  output = graph_data.get_vertice(input)
  puts "  Output value: #{output}"

  puts "\nremove_edge"
  input = ['A', 'B']
  puts "  Input value: #{input}"
  output = graph_data.remove_edge(*input)
  puts "  Output value: #{output}"
  puts "  Current data: #{graph_data.get}"

  puts "\nremove_edge"
  input = ['A', 'C']
  puts "  Input value: #{input}"
  output = graph_data.remove_edge(*input)
  puts "  Output value: #{output}"
  puts "  Current data: #{graph_data.get}"

  puts "\nget_edges"
  output = graph_data.get_edges
  puts "  Output value: #{output}"

  puts "\nremove_vertex"
  input = 'B'
  puts "  Input value: #{input}"
  output = graph_data.remove_vertex(input)
  puts "  Output value: #{output}"
  puts "  Current data: #{graph_data.get}"

  puts "\nremove_vertex"
  input = 'Z'
  puts "  Input value: #{input}"
  output = graph_data.remove_vertex(input)
  puts "  Output value: #{output}"
  puts "  Current data: #{graph_data.get}"

  puts "\nget_vertices"
  output = graph_data.get_vertices
  puts "  Output value: #{output}"

  puts "\nget_edges"
  output = graph_data.get_edges
  puts "  Output value: #{output}"

  puts "\nsize"
  output = graph_data.size
  puts "  Output value: #{output}"

  puts "\nget_vertice"
  input = 'B'
  puts "  Input value: #{input}"
  output = graph_data.get_vertice(input)
  puts "  Output value: #{output}"

  puts "\nclear"
  output = graph_data.clear
  puts "  Output value: #{output}"

  puts "\nis_empty?"
  output = graph_data.is_empty?
  puts "  Output value: #{output}"

  puts "\nsize"
  output = graph_data.size
  puts "  Output value: #{output}"

  puts "\nget_vertices"
  output = graph_data.get_vertices
  puts "  Output value: #{output}"

  puts "\nget_edges"
  output = graph_data.get_edges
  puts "  Output value: #{output}"

  puts "\nGraph TEST <----- end"
end

# Only run main if the script is run directly
main if __FILE__ == $0
